docs/articles/baseplatform/variant-concept-in-yocto.rst

```diff
@@ -0,0 +1,49 @@
+Variant concept in Yocto
+========================
+
+A Yocto project often has to support several variations in software and
+hardware composition, such as supporting a graphical software stack as well as
+a non-graphical software stack on several different CPU architectures.
+
+A variation in either hardware of software is in the Software Factory referred
+to as a *variant*.
+
+Each variant potentially needs its own set of Yocto configuration files (most
+notably, its own ``bblayers.conf`` and ``local.conf``). Yocto has a built-in
+way to handle deployment of configuration files with its ``TEMPLATECONF``
+variable which can be supplied to the ``oe-init-build-env`` script
+[#templateconf]_, and this scheme is also used in the Software Factory.
+
+.. _recommended-variant-management:
+
+Recommended variant management
+------------------------------
+
+The recommended way to handle variants is to keep one directory per variant in
+your layer. The variant directories should go into
+``<meta-yourlayer>/conf/<variant name>``. Each variant should supply a
```

Comment:
Actually, the current use (in meta-pelux) is to have `meta-pelux/meta-<hardware>-extras/conf-<variant>`, and I guess at least the hardware specific stuff makes sense to separate like that because one might want to include additional recipes. For meta-pelux, can we compromise and use `meta-pelux-meta-<hardware>-extras/conf/<variant>/` ?

Reply:
With this, the variants would be software variants, nested under hardware variants.

Reply:
For my current project, I have used a variant name with the format "hardwarevariant_softwarevariant" - an example would be "rpi3_nogfx" - this means I currently do not nest in the way you describe..

Reply:
Okay, but that would mean that any BSP specific recipes will be duplicated, right? If you have `rpi3_nogfx` and `rpi_qt` and both use some kernel bbappend?

Reply:
I think the selection of variants (for instance, maybe the project will only ever use one hardware) should be left up to the project - some projects may only need software variants.

Reply:
While I agree with that, we set a norm by writing in a certain way, and that way should be the "standard" way. That's why I'm picky about how we formulate it here.

Reply:
Well, recipes & bbappends are not necessarily grouped by variants, I suppose? For instance, a set of Qt-related bbappends would be used by all hardwares, and the recipe itself would contain MACHINE-specific sections.

My idea with these variants is mostly just for local.conf and bblayers.conf files.

Reply:
BTW, my #2 comment in this thread was not a reply to your comment #3.. It didn't show up in my UI.

Reply:
I guess my main point is that BSP != variant. You can have several variants using the same BSP.